File: server/rmhbox/state-sync.ts
/**
 * RMHbox — State Synchronization Service
 *
 * Manages heartbeat broadcasting, phase transition sync,
 * action sequence counters, and timer tick broadcasting.
 *
 * Responsibilities:
 * - Periodic heartbeat: sends full state snapshots to all in-game clients
 * - Phase transition sync: immediate full sync on lobby state changes
 * - Timer tick broadcasting: 1-second countdown ticks for timed phases
 *
 * Reference: docs/rmhbox/design-spec/core.md §8, §22
 * Implementation: docs/rmhbox/implementation/phase-3.md §3
 */

import { Server } from 'socket.io';
import { config } from './config';
import { LobbyManager } from './lobby-manager';
import { S2C } from '../../lib/rmhbox/events';
import type { RMHboxLobby } from './types';
import type { ClientLobbyState } from '../../lib/rmhbox/types';

/** Handle returned by startTimerBroadcast for cancel / pause / resume control. */
export interface TimerHandle {
  cancel: () => void;
  pause: () => void;
  resume: () => void;
  readonly isPaused: boolean;
}

export class StateSyncService {
  private readonly io: Server;
  private readonly lobbyManager: LobbyManager;
  private heartbeatInterval: ReturnType<typeof setInterval> | null = null;

  constructor(io: Server, lobbyManager: LobbyManager) {
    this.io = io;
    this.lobbyManager = lobbyManager;
  }

  // ─── Heartbeat (§3.2) ─────────────────────────────────────────

  startHeartbeat(): void {
    this.heartbeatInterval = setInterval(() => this.tick(), config.HEARTBEAT_INTERVAL_MS);
  }

  stopHeartbeat(): void {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }
  }

  /**
   * Heartbeat tick: sends per-player scoped state snapshots
   * to all connected members during active gameplay.
   */
  private tick(): void {
    for (const lobby of this.lobbyManager.getLobbies().values()) {
      if (lobby.state !== 'PLAYING') continue;
      this.sendFullSyncToAll(lobby);
    }
  }

  // ─── Phase Transition Sync (§3.3) ─────────────────────────────

  /**
   * Send a full ClientLobbyState snapshot to every connected member.
   * Called on every lobby state transition to ensure all clients
   * are fully synchronized at phase boundaries.
   */
  broadcastFullSync(lobbyId: string): void {
    const lobby = this.lobbyManager.getLobby(lobbyId);
    if (!lobby) return;
    this.sendFullSyncToAll(lobby);
  }

  /**
   * Build a client-safe state snapshot for a specific user.
   * Delegates to LobbyManager.buildClientState() which is the
   * ONLY exit point for state data — it strips internal fields
   * and scopes minigame state per player role.
   */
  buildClientState(lobby: RMHboxLobby, userId: string): ClientLobbyState {
    return this.lobbyManager.buildClientState(lobby, userId);
  }

  // ─── Timer Tick Broadcasting (§3.5) ────────────────────────────

  /**
   * Creates a 1-second countdown timer that broadcasts TIMER_TICK
   * actions to a lobby, then calls onComplete when done.
   *
   * First emits a TIMER_START action with `totalDuration` and `timeRemaining`
   * so the client header ring knows the full circle baseline. Then emits
   * TIMER_TICK every second with the decremented `timeRemaining`.
   *
   * @param lobbyId - The lobby to broadcast to
   * @param durationSeconds - Total seconds for the countdown
   * @param onComplete - Callback fired when countdown reaches 0
   * @returns A timer handle with cancel / pause / resume helpers
   */
  startTimerBroadcast(
    lobbyId: string,
    durationSeconds: number,
    onComplete: () => void,
  ): TimerHandle {
    let remaining = durationSeconds;
    let cancelled = false;
    let paused = false;
    let interval: ReturnType<typeof setInterval> | null = null;

    // Emit TIMER_START so the client knows totalDuration for ring animation
    this.lobbyManager.broadcastAction(lobbyId, {
      type: 'TIMER_START',
      payload: { totalDuration: durationSeconds, timeRemaining: remaining },
    });

    const tick = () => {
      if (cancelled || paused) return;

      remaining--;

      if (remaining < 0) {
        if (interval) clearInterval(interval);
        interval = null;
        if (!cancelled) {
          onComplete();
        }
        return;
      }

      // Broadcast TIMER_TICK action (decrement first so remaining matches display)
      this.lobbyManager.broadcastAction(lobbyId, {
        type: 'TIMER_TICK',
        payload: { timeRemaining: remaining },
      });
    };

    interval = setInterval(tick, 1000);

    const cancel = () => {
      cancelled = true;
      if (interval) { clearInterval(interval); interval = null; }
    };

    const pause = () => {
      if (paused || cancelled) return;
      paused = true;
      this.lobbyManager.broadcastAction(lobbyId, {
        type: 'TIMER_PAUSED',
        payload: { timeRemaining: Math.max(0, remaining) },
      });
    };

    const resume = () => {
      if (!paused || cancelled) return;
      paused = false;
      this.lobbyManager.broadcastAction(lobbyId, {
        type: 'TIMER_RESUMED',
        payload: { timeRemaining: Math.max(0, remaining) },
      });
    };

    return { cancel, pause, resume, get isPaused() { return paused; } };
  }

  /**
   * Broadcast an "infinite" timer that shows a full ring with an ∞ icon.
   * No ticking, no auto-complete. The host advances manually via force-skip.
   * Sentinel value: `totalDuration: -1, timeRemaining: -1`.
   * Always shows the host "Next" button (`showSkip: true`).
   */
  broadcastInfiniteTimer(lobbyId: string): TimerHandle {
    let cancelled = false;
    let paused = false;

    this.lobbyManager.broadcastAction(lobbyId, {
      type: 'TIMER_START',
      payload: { totalDuration: -1, timeRemaining: -1, showSkip: true },
    });

    const cancel = () => { cancelled = true; };

    const pause = () => {
      if (paused || cancelled) return;
      paused = true;
      this.lobbyManager.broadcastAction(lobbyId, {
        type: 'TIMER_PAUSED',
        payload: { timeRemaining: -1 },
      });
    };

    const resume = () => {
      if (!paused || cancelled) return;
      paused = false;
      this.lobbyManager.broadcastAction(lobbyId, {
        type: 'TIMER_RESUMED',
        payload: { timeRemaining: -1 },
      });
    };

    return { cancel, pause, resume, get isPaused() { return paused; } };
  }

  // ─── Internal Helpers ─────────────────────────────────────────

  /**
   * Send per-player scoped state snapshots to all connected
   * players and spectators in a lobby.
   */
  private sendFullSyncToAll(lobby: RMHboxLobby): void {
    // Send per-player scoped state snapshots
    for (const [userId, player] of lobby.players) {
      if (player.isConnected && player.socketId) {
        const clientState = this.buildClientState(lobby, userId);
        this.io
          .to(`lobby:${lobby.id}:player:${userId}`)
          .emit(S2C.LOBBY_STATE_SNAPSHOT, clientState);
      }
    }

    // Spectators get the spectator view
    for (const [userId, spectator] of lobby.spectators) {
      if (spectator.isConnected && spectator.socketId) {
        const clientState = this.buildClientState(lobby, userId);
        this.io
          .to(`lobby:${lobby.id}:player:${userId}`)
          .emit(S2C.LOBBY_STATE_SNAPSHOT, clientState);
      }
    }
  }
}

File: go-services/internal/rmhbox/statesync.go
package rmhbox

import (
	"sync"
	"time"
)

// ─── State synchronization (state-sync.ts) ───────────────────────────────────
//
// Two responsibilities, faithfully ported:
//   - a 10s heartbeat that pushes per-player scoped full snapshots to all
//     members of every PLAYING lobby;
//   - per-lobby, pausable 1s TIMER_TICK broadcasts driving the phase rings.

// TimerHandle controls a running lifecycle timer broadcast.
type TimerHandle struct {
	mu        sync.Mutex
	cancelled bool
	paused    bool
	remaining int
	stop      chan struct{}
	lm        *LobbyManager
	lobbyID   string
	infinite  bool
}

// IsPaused reports the paused state.
func (t *TimerHandle) IsPaused() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.paused
}

// Cancel stops the timer broadcast.
func (t *TimerHandle) Cancel() {
	t.mu.Lock()
	if t.cancelled {
		t.mu.Unlock()
		return
	}
	t.cancelled = true
	// Closing the channel is the signal; the `cancelled` guard above makes this
	// close-once. We deliberately do NOT nil t.stop — the broadcast goroutine
	// reads it without the lock, so mutating it here would be a data race.
	if t.stop != nil {
		close(t.stop)
	}
	t.mu.Unlock()
}

// Pause suspends ticking and emits TIMER_PAUSED.
func (t *TimerHandle) Pause() {
	t.mu.Lock()
	if t.paused || t.cancelled {
		t.mu.Unlock()
		return
	}
	t.paused = true
	rem := t.remaining
	t.mu.Unlock()
	if t.infinite {
		rem = -1
	} else if rem < 0 {
		rem = 0
	}
	t.lm.broadcastAction(t.lobbyID, "TIMER_PAUSED", map[string]any{"timeRemaining": rem})
}

// Resume restarts ticking and emits TIMER_RESUMED.
func (t *TimerHandle) Resume() {
	t.mu.Lock()
	if !t.paused || t.cancelled {
		t.mu.Unlock()
		return
	}
	t.paused = false
	rem := t.remaining
	t.mu.Unlock()
	if t.infinite {
		rem = -1
	} else if rem < 0 {
		rem = 0
	}
	t.lm.broadcastAction(t.lobbyID, "TIMER_RESUMED", map[string]any{"timeRemaining": rem})
}

// StateSync drives heartbeats and lifecycle timer broadcasts.
type StateSync struct {
	lm     *LobbyManager
	hbStop chan struct{}
}

// NewStateSync constructs the state-sync service.
func NewStateSync(lm *LobbyManager) *StateSync { return &StateSync{lm: lm} }

func (s *StateSync) startHeartbeat() {
	s.hbStop = make(chan struct{})
	go func() {
		t := time.NewTicker(heartbeatInterval)
		defer t.Stop()
		for {
			select {
			case <-s.hbStop:
				return
			case <-t.C:
				s.tick()
			}
		}
	}()
}

func (s *StateSync) stopHeartbeat() {
	if s.hbStop != nil {
		close(s.hbStop)
	}
}

func (s *StateSync) tick() {
	for _, l := range s.lm.Lobbies() {
		l.mu.Lock()
		playing := l.State == StatePlaying
		l.mu.Unlock()
		if playing {
			s.sendFullSyncToAll(l)
		}
	}
}

// broadcastFullSync pushes per-recipient snapshots on phase boundaries.
func (s *StateSync) broadcastFullSync(lobbyID string) {
	l := s.lm.GetLobby(lobbyID)
	if l == nil {
		return
	}
	s.sendFullSyncToAll(l)
}

func (s *StateSync) sendFullSyncToAll(l *Lobby) {
	l.mu.Lock()
	playerIDs := make([]string, 0, len(l.Players))
	for id, p := range l.Players {
		if p.IsConnected && p.ConnID != "" {
			playerIDs = append(playerIDs, id)
		}
	}
	specIDs := make([]string, 0, len(l.Spectators))
	for id, sp := range l.Spectators {
		if sp.IsConnected && sp.ConnID != "" {
			specIDs = append(specIDs, id)
		}
	}
	l.mu.Unlock()

	for _, id := range playerIDs {
		s.lm.sendToUser(id, "rmhbox:lobby:state_snapshot", s.lm.buildClientState(l, id))
	}
	for _, id := range specIDs {
		s.lm.sendToUser(id, "rmhbox:lobby:state_snapshot", s.lm.buildClientState(l, id))
	}
}

// startTimerBroadcast emits TIMER_START then a TIMER_TICK every second; onComplete
// fires when the countdown crosses below zero. Pausable.
func (s *StateSync) startTimerBroadcast(lobbyID string, durationSeconds int, onComplete func()) *TimerHandle {
	h := &TimerHandle{remaining: durationSeconds, stop: make(chan struct{}), lm: s.lm, lobbyID: lobbyID}
	s.lm.broadcastAction(lobbyID, "TIMER_START", map[string]any{"totalDuration": durationSeconds, "timeRemaining": durationSeconds})

	// Capture the stop channel once: it is set above before this goroutine
	// starts (happens-before via goroutine creation) and never reassigned, so
	// reading it here needs no lock.
	stop := h.stop
	go func() {
		t := time.NewTicker(time.Second)
		defer t.Stop()
		for {
			select {
			case <-stop:
				return
			case <-t.C:
				h.mu.Lock()
				if h.cancelled || h.paused {
					h.mu.Unlock()
					continue
				}
				h.remaining--
				rem := h.remaining
				cancelled := h.cancelled
				h.mu.Unlock()
				if rem < 0 {
					if !cancelled && onComplete != nil {
						onComplete()
					}
					return
				}
				s.lm.broadcastAction(lobbyID, "TIMER_TICK", map[string]any{"timeRemaining": rem})
			}
		}
	}()
	return h
}

// broadcastInfiniteTimer shows a full ring with no ticking (host advances).
func (s *StateSync) broadcastInfiniteTimer(lobbyID string) *TimerHandle {
	h := &TimerHandle{remaining: -1, stop: make(chan struct{}), lm: s.lm, lobbyID: lobbyID, infinite: true}
	s.lm.broadcastAction(lobbyID, "TIMER_START", map[string]any{"totalDuration": -1, "timeRemaining": -1, "showSkip": true})
	return h
}
